File: sections/66_master_math_mixed_contest/problems/d_weighted_compositions/solution.cpp
#include <bits/stdc++.h>
#include "../../../65_polynomial_algorithms_recurrences/ntt.hpp"
using namespace std;
using course_ntt::convolution;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, target;
    if (!(cin >> n >> target)) return 0;
    struct Compare {
        bool operator()(const vector<int> &a, const vector<int> &b) const {
            return a.size() > b.size();
        }
    };
    priority_queue<vector<int>, vector<vector<int>>, Compare> heap;
    for (int i = 0, weight, cap; i < n; ++i) {
        cin >> weight >> cap;
        int maximum = static_cast<int>(min(1LL * target, 1LL * weight * cap));
        vector<int> polynomial(maximum + 1);
        for (int value = 0; value <= maximum; value += weight) polynomial[value] = 1;
        if (polynomial.size() > 1) heap.push(move(polynomial));
    }
    if (heap.empty()) {
        cout << (target == 0) << '\n';
        return 0;
    }
    while (heap.size() > 1) {
        auto a = heap.top(); heap.pop();
        auto b = heap.top(); heap.pop();
        heap.push(convolution(a, b, target + 1));
    }
    const auto &answer = heap.top();
    cout << (target < static_cast<int>(answer.size()) ? answer[target] : 0) << '\n';
}

File: sections/66_master_math_mixed_contest/problems/d_weighted_compositions/solution.py
import heapq
import sys
from pathlib import Path

SECTION_65 = Path(__file__).resolve().parents[3] / "65_polynomial_algorithms_recurrences"
sys.path.insert(0, str(SECTION_65))
from ntt import convolution


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, target = data[:2]
    heap: list[tuple[int, int, list[int]]] = []
    index = 2
    serial = 0
    for _ in range(n):
        weight, cap = data[index : index + 2]
        index += 2
        maximum = min(target, weight * cap)
        polynomial = [0] * (maximum + 1)
        for value in range(0, maximum + 1, weight):
            polynomial[value] = 1
        if len(polynomial) > 1:
            heapq.heappush(heap, (len(polynomial), serial, polynomial))
            serial += 1
    if not heap:
        print(int(target == 0))
        return
    while len(heap) > 1:
        _, _, left = heapq.heappop(heap)
        _, _, right = heapq.heappop(heap)
        product = convolution(left, right, target + 1)
        heapq.heappush(heap, (len(product), serial, product))
        serial += 1
    answer = heap[0][2]
    print(answer[target] if target < len(answer) else 0)


if __name__ == "__main__":
    main()
